compiler-core/src/type_.rs

```diff
@@ -280,7 +280,12 @@ impl Type {
                 inferred_variant, ..
             } => *inferred_variant = None,
             Type::Var { type_ } => type_.borrow_mut().generalise_custom_type_variant(),
-            Type::Fn { .. } | Type::Tuple { .. } => {}
+            Type::Tuple { elems } => {
+                for element in elems {
+                    Arc::make_mut(element).generalise_custom_type_variant();
```

Comment:
I don't really like how this has to clone every tuple element, even if they don't need to be modified at all. I can't really find a better way to do it though